### PR TITLE
check for holding requests

### PR DESCRIPTION
**Pull request by tootedom**
# Issue Type

- Feature Pull Request
- Bugfix Pull Request

## Summary

New branch as per the discussion on #222 

Hopefully this deals with the merge conflict issue.  Also easier to see the changes, and less files modified.

Let me know.  Sorry for the back and forth hassle.

## Code contribution checklist

1. [x] The contribution fixes a single existing github issue, and it is linked
   to it.
1. [x] The code is as simple as possible, readable and follows the idiomatic Go
  [guidelines](https://golang.org/doc/effective_go.html).
1. [ ] All new functionality is covered by automated test cases so the overall
  test coverage doesn't decrease.
1. [ ] No issues are reported when running `make full-test`.
1. [ ] The logs should be kept as clean as possible (use log levels as
  appropriate) and formatted consistently to the existing log output.
1. [ ] A code reviewer reproduced the problem and can confirm the code
  contribution actually resolves it.

### CODE REVIEW

**Comment by tootedom** [comment]
The slight decrease in the coverage in core/autoscaling.go is the missing tests on `havingReadyToAttachSpotInstance()` method.  This I'll be looking to address in the future not use long live SIR requests that I'll work on (currently working on it...).

**Comment by tootedom** [comment]
Hola,
I've been running the autospotting for a week on this branch.  I've not had any issue with it.  However, it's never got to the scenario where AWS has put the request in holding.

I can see the branch is out of date with master.  I'll update and push this weekend.  

Not sure how you want to proceed.  Am happy to keep it running for another week (after the master update) if needs.

(I've been looking into the cancelling the spot request after 4 mins.  So far progress is ok.  I'll hopefully get some time next week in the evenings to work on it and get an initial WIP pull to you).

cheers
/dom

**Comment by tootedom** [comment]
Hi there,

Noticed that we hit one of the amazon holding request issues today (and late yesterday)

```
21:34:19
2018/03/12 21:34:19 autoscaling.go:553: asg-qa-rap_0-0-2-xxx_ami-xxxxx Open bid found for current AutoScaling Group, waiting for the instance to start so it can be tagged...

21:34:19
2018/03/12 21:34:19 autoscaling.go:519: asg-qa-rap_0-0-2-xxx_ami-xxxxx5 Spot Request (sir-xyzhjur) is in holding by Amazon:capacity-oversubscribed

21:34:19
2018/03/12 21:34:19 autoscaling.go:639: asg-qa-rap_0-0-2-xxx_ami-xxxxx No active unfulfilled bid was found

21:34:19
2018/03/12 21:34:19 autoscaling.go:308: Waiting for next run while processing asg-qa-rap_0-0-2-xxx_ami-xxxxx
```

The code lambda did not wait on the spot instance request, and basically left it until the next run.  It took 4 runs (40 mins), until Amazon had the capacity available.  The spot request was still open during that time, and the code ignored it (waited for the Amazon to fulfill the request).

Here's the code handling another of the holding request states

```
2018/03/13 15:38:25 autoscaling.go:519: asg-qa-epp_0-1-0-xxx_ami-xxxxx Spot Request (sir-46kkekjd) is in holding by Amazon:capacity-not-available
```

----

I notice the pull is out of date again.  I'll update and push a merge with master.  Hopefully the above gives you some good positive vibes on merging :-)

I should get some time at the weekend to get you the short lived spot request pull (sorry for delay).

File: core/spot_instance_request.go [hunk 1]
```diff
@@ -19,6 +19,39 @@ type spotInstanceRequest struct {
...
+	return s.State != nil && *s.State == "open"
+}
+
+// Before waiting for an instance we check if amazon has put the open request
+// in a holding state (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/spot-bid-status.html).
+func (s *spotInstanceRequest) isHoldingRequest() bool {
+	return s.isRequestOpen() && s.Status != nil && s.Status.Code != nil && hasHoldingRequestStatus(*s.Status.Code)
+}
+
+func hasHoldingRequestStatus(code string) bool {
```
Comment (cristim) [inline core/spot_instance_request.go]: A small nitpick:

We can merge these case statements like this: https://github.com/golang/go/wiki/Switch#multiple-cases

Comment (tootedom) [reply core/spot_instance_request.go]: I'll have a look to see if I can combine the branches better.  Not sure if multiple-cases is possible, as it's not switching on the single value and check it.  It's different "switch" values.  I can probably pass `s.Status` to `hasHoldingRequestStatus(..)` and let it do it's thing.

Will look when I update the to master tonight. 
cheers
/dom

**Comment by tootedom** [comment]
Have update the nil check and moved to method to make easier to read.
Let me know if that's better.
cheers
/dom

**Comment by tootedom** [comment]
I'm getting close on the short live SIR request.  So you might want to hold off on merging this.

**Comment by cristim** [comment]
Cool, let's close this and please open a PR for the short lived SIRs so we can review it while in progress

**Pull request closed without merging**